File: protocol/impls/java/AttribCurve.java
import java.math.BigInteger;
import java.security.MessageDigest;
import java.util.Arrays;

// §4 Strategy B — the pinned ECDSA curve-vector set (`sm attrib-curve`).
//
// Mirrors impls/c/src/attrib_curve.c (attrib_cmd_curve) + attrib.c (attrib_real_endtoend)
// byte-for-byte. Runs Java's OWN secp256k1 (Secp.java) and must print output identical
// to the C reference. Covers pinned P/N/N_HALF constants, on-curve membership at the
// edges, ECDSA verify accept/reject at the scalar boundaries, RFC-6979 deterministic
// (r,s) + canonical-DER known-answers, the tiny-key KAT, the PRIMARY `combined` digest,
// and the real-curve end-to-end attribution vectors + `combined_e2e`.
final class AttribCurve {

    // a streaming SHA-256 accumulator (the C single-CTX feed)
    private static MessageDigest sha256ctx() {
        try { return MessageDigest.getInstance("SHA-256"); }
        catch (Exception e) { throw new RuntimeException(e); }
    }

    private static final byte[] CV_P     = Secp.be32(Secp.P);
    private static final byte[] CV_N     = Secp.be32(Secp.N);
    private static final byte[] CV_NHALF = Secp.be32(Secp.N_HALF);
    private static final byte[] CV_GX    = Secp.be32(Secp.GX);
    private static final byte[] CV_GY    = Secp.be32(Secp.GY);

    static void run() {
        StringBuilder out = new StringBuilder();
        MessageDigest comb = sha256ctx();

        // ── 1. pinned constants ──────────────────────────────────────────────
        out.append("p ").append(Hex.enc(CV_P)).append("\n");         comb.update(CV_P);
        out.append("n ").append(Hex.enc(CV_N)).append("\n");         comb.update(CV_N);
        out.append("nhalf ").append(Hex.enc(CV_NHALF)).append("\n"); comb.update(CV_NHALF);

        // ── 2. on-curve membership at the edges ──────────────────────────────
        String[] ocName = new String[16];
        byte[][] ocKey = new byte[16][];
        int noc = 0;
        byte[] buf;
        // G uncompressed (on)
        buf = new byte[65]; buf[0] = 0x04; sys(buf, 1, CV_GX); sys(buf, 33, CV_GY);
        ocKey[noc] = buf; ocName[noc] = "oc_G_uncomp"; noc++;
        // G compressed even (on)
        buf = new byte[33]; buf[0] = 0x02; sys(buf, 1, CV_GX);
        ocKey[noc] = buf; ocName[noc] = "oc_G_comp02"; noc++;
        // G compressed odd-prefix (still on curve: yields p−Gy)
        buf = new byte[33]; buf[0] = 0x03; sys(buf, 1, CV_GX);
        ocKey[noc] = buf; ocName[noc] = "oc_G_comp03"; noc++;
        // (Gx, Gy^lsb) uncompressed (off curve)
        buf = new byte[65]; buf[0] = 0x04; sys(buf, 1, CV_GX); sys(buf, 33, CV_GY); buf[64] ^= 0x01;
        ocKey[noc] = buf; ocName[noc] = "oc_G_badY"; noc++;
        // compressed X=0
        buf = new byte[33]; buf[0] = 0x02;
        ocKey[noc] = buf; ocName[noc] = "oc_X0"; noc++;
        // compressed X=1
        buf = new byte[33]; buf[0] = 0x02; buf[32] = 1;
        ocKey[noc] = buf; ocName[noc] = "oc_X1"; noc++;
        // uncompressed X=p ⇒ decode-reject
        buf = new byte[65]; buf[0] = 0x04; sys(buf, 1, CV_P); sys(buf, 33, CV_GY);
        ocKey[noc] = buf; ocName[noc] = "oc_Xeqp"; noc++;
        // compressed X=p ⇒ decode-reject
        buf = new byte[33]; buf[0] = 0x02; sys(buf, 1, CV_P);
        ocKey[noc] = buf; ocName[noc] = "oc_comp_Xeqp"; noc++;
        // bad prefix 0x05 ⇒ reject
        buf = new byte[33]; buf[0] = 0x05; sys(buf, 1, CV_GX);
        ocKey[noc] = buf; ocName[noc] = "oc_badprefix"; noc++;
        for (int i = 0; i < noc; i++) {
            int v = Secp.onCurve(ocKey[i]) ? 1 : 0;
            out.append(ocName[i]).append(" ").append(v).append("\n");
            comb.update((byte) v); comb.update(ocKey[i]);
        }

        // ── 3 & 4. RFC-6979 sign + ECDSA verify at the boundaries ────────────
        for (int i = 0; i < 4; i++) {
            byte[] priv = new byte[32];
            priv[28] = (byte) 0xC0; priv[29] = (byte) 0xFF; priv[30] = (byte) 0xEE; priv[31] = (byte) (0x10 + i);
            byte[] pub = Secp.pubkey(priv);
            if (pub == null) { out.append("sig").append(i).append(" PUBFAIL\n"); continue; }
            byte[] m = Sm.utf8("strategy-b curve vector " + i);
            byte[] h = Hashes.sha256(m);
            byte[][] rs = Secp.ecdsaSign(priv, h);
            if (rs == null) { out.append("sig").append(i).append(" SIGNFAIL\n"); continue; }
            byte[] r = rs[0], s = rs[1];
            byte[] der = Secp.derSig(r, s);
            out.append("sig").append(i)
               .append(" pub=").append(Hex.enc(pub))
               .append(" r=").append(Hex.enc(r))
               .append(" s=").append(Hex.enc(s))
               .append(" der=").append(Hex.enc(der)).append("\n");
            comb.update(pub); comb.update(r); comb.update(s); comb.update(der);

            // verify boundary battery
            byte[] zero = new byte[32];
            byte[] hbad = Arrays.copyOf(h, 32); hbad[0] ^= 0x01;
            byte[] hiS = new byte[32];                       // high-S = n - s (byte subtraction)
            { int borrow = 0;
              for (int k = 31; k >= 0; k--) {
                  int d = (CV_N[k] & 0xFF) - (s[k] & 0xFF) - borrow;
                  if (d < 0) { d += 256; borrow = 1; } else borrow = 0;
                  hiS[k] = (byte) d;
              } }
            byte[] wrongpub = Arrays.copyOf(pub, 33); wrongpub[0] ^= 0x01;
            String[] vn = { "valid", "tamper", "r0", "s0", "rN", "sN", "highS", "wrongpk" };
            byte[][] vh = { h, hbad, h, h, h, h, h, h };
            byte[][] vr = { r, r, zero, r, CV_N, r, r, r };
            byte[][] vs = { s, s, s, zero, s, CV_N, hiS, s };
            byte[][] vp = { pub, pub, pub, pub, pub, pub, pub, wrongpub };
            out.append("ver").append(i);
            for (int t = 0; t < vn.length; t++) {
                int v = Secp.ecdsaVerify(vh[t], vr[t], vs[t], vp[t]) ? 1 : 0;
                out.append(" ").append(vn[t]).append("=").append(v);
                comb.update((byte) v);
            }
            out.append("\n");
        }

        // ── 5. tiny-key KAT: priv=1 ⇒ pub=G ; priv=2 ⇒ pub=2G ─────────────────
        { byte[] p1 = new byte[32]; p1[31] = 1; byte[] pk1 = Secp.pubkey(p1);
          out.append("priv1_pub=").append(Hex.enc(pk1)).append("\n"); comb.update(pk1);
          byte[] p2 = new byte[32]; p2[31] = 2; byte[] pk2 = Secp.pubkey(p2);
          out.append("priv2_pub=").append(Hex.enc(pk2)).append("\n"); comb.update(pk2); }

        // PRIMARY cross-language digest (sections 1–5)
        byte[] cd = comb.digest();
        out.append("combined ").append(Hex.enc(cd)).append("\n");

        // ── 6. end-to-end: real legacy sighash signing + attribute() ─────────
        MessageDigest e2e = sha256ctx();
        realEndToEnd(out, e2e);
        byte[] ed = e2e.digest();
        out.append("combined_e2e ").append(Hex.enc(ed)).append("\n");

        System.out.print(out);
    }

    // ── shared tx skeleton: 1 input (outpoint 0x11.. seq=FFFFFFFF), 1 output
    //    (value=100000, OP_RETURN), version 1, locktime 0. attribute()'s
    //    recomputed sighash equals the one we sign. ──────────────────────────
    private static Attrib.Tx skeleton(byte[] scriptSig) {
        Attrib.Tx t = new Attrib.Tx(); t.version = 1; t.locktime = 0;
        Attrib.In in = new Attrib.In();
        // C outpoint is 36 bytes all 0x11: txid[32]=0x11.., vout=0x11111111.
        in.prevTxid = new byte[32]; Arrays.fill(in.prevTxid, (byte) 0x11);
        in.prevVout = 0x11111111L;
        in.sequence = 0xFFFFFFFFL;
        in.scriptSig = scriptSig;
        t.ins.add(in);
        Attrib.Out o = new Attrib.Out();
        o.value = BigInteger.valueOf(100000); o.spk = new byte[]{0x6a};
        t.outs.add(o);
        return t;
    }
    // legacy sighash of the skeleton with the given scriptCode for input 0
    private static byte[] skelSighash(byte[] scriptCode) {
        Attrib.Tx sk = skeleton(new byte[0]);
        return Attrib.sighash(sk, 0, scriptCode);
    }

    private static void emitE2E(StringBuilder out, MessageDigest comb, String name, Attrib.Tx t) {
        Attrib.Result res = Attrib.attribute(t, 0);
        out.append(name).append(" ").append(res.status).append(":").append(Hex.enc(res.identity)).append("\n");
        comb.update((byte) res.status);
        comb.update(res.sighash);
        comb.update(res.identity);
    }

    static void realEndToEnd(StringBuilder out, MessageDigest comb) {
        Attrib.realCurve = true;
        try {
            // ── A. P2PKH, correctly signed ⇒ FOUND ───────────────────────────
            {
                byte[] priv = new byte[32]; priv[31] = 0x2A;     // priv = 42
                byte[] pub = Secp.pubkey(priv);
                byte[] h160 = Hashes.hash160(pub);
                byte[] sc = Attrib.p2pkhScript(h160);            // OP_DUP OP_HASH160 <20> OP_EQUALVERIFY OP_CHECKSIG
                byte[] sh = skelSighash(sc);
                byte[][] rs = Secp.ecdsaSign(priv, sh);
                byte[] der = Secp.derSig(rs[0], rs[1]);
                byte[] ss = Fold.concat(Attrib.pushOf(der), Attrib.pushOf(pub));
                emitE2E(out, comb, "e2e_p2pkh_valid", skeleton(ss));
            }
            // ── B. P2PKH, WRONG key ⇒ verify-drop (status 2) ──────────────────
            {
                byte[] priv = new byte[32]; priv[31] = 0x2A;
                byte[] wrong = new byte[32]; wrong[31] = 0x2B;
                byte[] pub = Secp.pubkey(priv);
                byte[] h160 = Hashes.hash160(pub);
                byte[] sc = Attrib.p2pkhScript(h160);
                byte[] sh = skelSighash(sc);
                byte[][] rs = Secp.ecdsaSign(wrong, sh);          // wrong key signs
                byte[] der = Secp.derSig(rs[0], rs[1]);
                byte[] ss = Fold.concat(Attrib.pushOf(der), Attrib.pushOf(pub));
                emitE2E(out, comb, "e2e_p2pkh_wrongkey", skeleton(ss));
            }
            // ── C. 2-of-2 P2SH multisig, two correct in-order sigs ⇒ FOUND ────
            {
                byte[][] priv = new byte[2][32], pub = new byte[2][];
                for (int i = 0; i < 2; i++) { priv[i][31] = (byte) (0x50 + i); pub[i] = Secp.pubkey(priv[i]); }
                Buf rsB = new Buf();
                rsB.u8(0x52);                                     // OP_2
                for (int i = 0; i < 2; i++) rsB.u8(0x21).bytes(pub[i]);
                rsB.u8(0x52).u8(0xAE);                            // OP_2 OP_CHECKMULTISIG
                byte[] redeem = rsB.toBytes();
                byte[] sh = skelSighash(redeem);
                Buf ssB = new Buf(); ssB.u8(0x00);                // NULLDUMMY
                for (int i = 0; i < 2; i++) {
                    byte[][] rs = Secp.ecdsaSign(priv[i], sh);
                    ssB.bytes(Attrib.pushOf(Secp.derSig(rs[0], rs[1])));
                }
                ssB.bytes(Attrib.pushOf(redeem));
                emitE2E(out, comb, "e2e_multisig_valid", skeleton(ssB.toBytes()));
            }
        } finally {
            Attrib.realCurve = false;
        }
    }

    private static void sys(byte[] dst, int off, byte[] src) { System.arraycopy(src, 0, dst, off, src.length); }
}

File: protocol/impls/rust/src/attrib_curve.rs
//! §4 Strategy B — the pinned ECDSA curve-vector set (`sm attrib-curve`).
//!
//! Faithful port of impls/c/src/attrib_curve.c (`attrib_cmd_curve`) + the end-to-end
//! pipeline in attrib.c (`attrib_real_endtoend`). Prints output byte-identical to the
//! C reference: pinned P/N/N_HALF constants, on-curve edge vectors, RFC-6979
//! deterministic (r,s)+canonical-DER, the verify boundary battery, the tiny-key KAT,
//! the PRIMARY `combined` digest, then the real-curve end-to-end vectors + `combined_e2e`.

use crate::attrib::{attribute, legacy_sighash, parse_tx, set_real_curve, ParsedTx, TxIn, TxOut};
use crate::ripemd160::hash160;
use crate::secp256k1::{secp_ecdsa_sign, secp_ecdsa_verify, secp_on_curve, secp_pubkey};
use crate::sha256::Sha256;

const HEXD: &[u8; 16] = b"0123456789abcdef";
fn puthex(out: &mut String, d: &[u8]) {
    for &b in d {
        out.push(HEXD[(b >> 4) as usize] as char);
        out.push(HEXD[(b & 15) as usize] as char);
    }
}
fn sha(d: &[u8]) -> [u8; 32] {
    let mut c = Sha256::new();
    c.update(d);
    c.finalize()
}

// secp256k1 p, n, n/2, G — big-endian constants (mirror secp256k1.c).
const CV_P: [u8; 32] = [
    0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF,
    0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFE, 0xFF, 0xFF, 0xFC, 0x2F,
];
const CV_N: [u8; 32] = [
    0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFE,
    0xBA, 0xAE, 0xDC, 0xE6, 0xAF, 0x48, 0xA0, 0x3B, 0xBF, 0xD2, 0x5E, 0x8C, 0xD0, 0x36, 0x41, 0x41,
];
const CV_NHALF: [u8; 32] = [
    0x7F, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF,
    0x5D, 0x57, 0x6E, 0x73, 0x57, 0xA4, 0x50, 0x1D, 0xDF, 0xE9, 0x2F, 0x46, 0x68, 0x1B, 0x20, 0xA0,
];
const CV_GX: [u8; 32] = [
    0x79, 0xBE, 0x66, 0x7E, 0xF9, 0xDC, 0xBB, 0xAC, 0x55, 0xA0, 0x62, 0x95, 0xCE, 0x87, 0x0B, 0x07,
    0x02, 0x9B, 0xFC, 0xDB, 0x2D, 0xCE, 0x28, 0xD9, 0x59, 0xF2, 0x81, 0x5B, 0x16, 0xF8, 0x17, 0x98,
];
const CV_GY: [u8; 32] = [
    0x48, 0x3A, 0xDA, 0x77, 0x26, 0xA3, 0xC4, 0x65, 0x5D, 0xA4, 0xFB, 0xFC, 0x0E, 0x11, 0x08, 0xA8,
    0xFD, 0x17, 0xB4, 0x48, 0xA6, 0x85, 0x54, 0x19, 0x9C, 0x47, 0xD0, 0x8F, 0xFB, 0x10, 0xD4, 0xB8,
];

// canonical strict-DER (r,s) ‖ SIGHASH_ALL.
fn der_int(out: &mut Vec<u8>, v: &[u8; 32]) {
    let mut i = 0usize;
    while i < 31 && v[i] == 0 {
        i += 1;
    }
    let len = 32 - i;
    let pad = if v[i] & 0x80 != 0 { 1 } else { 0 };
    out.push(0x02);
    out.push((len + pad) as u8);
    if pad == 1 {
        out.push(0x00);
    }
    out.extend_from_slice(&v[i..]);
}
fn der_sig(r: &[u8; 32], s: &[u8; 32]) -> Vec<u8> {
    let mut body = Vec::new();
    der_int(&mut body, r);
    der_int(&mut body, s);
    let mut out = Vec::with_capacity(body.len() + 3);
    out.push(0x30);
    out.push(body.len() as u8);
    out.extend_from_slice(&body);
    out.push(0x01); // SIGHASH_ALL
    out
}

pub fn run() -> i32 {
    let mut comb = Sha256::new();
    let mut line = String::new();

    // ── 1. pinned constants ────────────────────────────────────────────────────
    line.clear();
    line.push_str("p ");
    puthex(&mut line, &CV_P);
    println!("{}", line);
    comb.update(&CV_P);
    line.clear();
    line.push_str("n ");
    puthex(&mut line, &CV_N);
    println!("{}", line);
    comb.update(&CV_N);
    line.clear();
    line.push_str("nhalf ");
    puthex(&mut line, &CV_NHALF);
    println!("{}", line);
    comb.update(&CV_NHALF);

    // ── 2. on-curve membership at the edges ────────────────────────────────────
    let mut oc: Vec<(&str, Vec<u8>)> = Vec::new();
    // G uncompressed (on)
    {
        let mut b = vec![0x04u8];
        b.extend_from_slice(&CV_GX);
        b.extend_from_slice(&CV_GY);
        oc.push(("oc_G_uncomp", b));
    }
    // G compressed even (on)
    {
        let mut b = vec![0x02u8];
        b.extend_from_slice(&CV_GX);
        oc.push(("oc_G_comp02", b));
    }
    // G compressed odd-prefix (still on curve)
    {
        let mut b = vec![0x03u8];
        b.extend_from_slice(&CV_GX);
        oc.push(("oc_G_comp03", b));
    }
    // (Gx, Gy^lsb) uncompressed (off curve)
    {
        let mut b = vec![0x04u8];
        b.extend_from_slice(&CV_GX);
        b.extend_from_slice(&CV_GY);
        let l = b.len();
        b[l - 1] ^= 0x01;
        oc.push(("oc_G_badY", b));
    }
    // compressed X=0
    {
        let mut b = vec![0x02u8];
        b.extend_from_slice(&[0u8; 32]);
        oc.push(("oc_X0", b));
    }
    // compressed X=1
    {
        let mut b = vec![0x02u8];
        let mut x = [0u8; 32];
        x[31] = 1;
        b.extend_from_slice(&x);
        oc.push(("oc_X1", b));
    }
    // uncompressed X>=p (X = p)
    {
        let mut b = vec![0x04u8];
        b.extend_from_slice(&CV_P);
        b.extend_from_slice(&CV_GY);
        oc.push(("oc_Xeqp", b));
    }
    // compressed X>=p
    {
        let mut b = vec![0x02u8];
        b.extend_from_slice(&CV_P);
        oc.push(("oc_comp_Xeqp", b));
    }
    // bad prefix 0x05
    {
        let mut b = vec![0x05u8];
        b.extend_from_slice(&CV_GX);
        oc.push(("oc_badprefix", b));
    }
    for (name, key) in &oc {
        let v = secp_on_curve(key) as i32;
        println!("{} {}", name, v);
        comb.update(&[v as u8]);
        comb.update(key);
    }

    // ── 3 & 4. RFC-6979 deterministic sign + ECDSA verify at the boundaries ─────
    for i in 0..4 {
        let mut priv_ = [0u8; 32];
        priv_[28] = 0xC0;
        priv_[29] = 0xFF;
        priv_[30] = 0xEE;
        priv_[31] = 0x10 + i as u8;
        let pubk = match secp_pubkey(&priv_) {
            Some(p) => p,
            None => {
                println!("sig{} PUBFAIL", i);
                continue;
            }
        };
        let m = format!("strategy-b curve vector {}", i);
        let h = sha(m.as_bytes());
        let (r, s) = match secp_ecdsa_sign(&priv_, &h) {
            Some(rs) => rs,
            None => {
                println!("sig{} SIGNFAIL", i);
                continue;
            }
        };
        let der = der_sig(&r, &s);
        line.clear();
        line.push_str(&format!("sig{} pub=", i));
        puthex(&mut line, &pubk);
        line.push_str(" r=");
        puthex(&mut line, &r);
        line.push_str(" s=");
        puthex(&mut line, &s);
        line.push_str(" der=");
        puthex(&mut line, &der);
        println!("{}", line);
        comb.update(&pubk);
        comb.update(&r);
        comb.update(&s);
        comb.update(&der);

        // verify boundary battery
        let zero = [0u8; 32];
        let mut hbad = h;
        hbad[0] ^= 0x01;
        // high-S = n - s (byte subtraction)
        let mut his = [0u8; 32];
        {
            let mut borrow: i32 = 0;
            for k in (0..32).rev() {
                let mut d = CV_N[k] as i32 - s[k] as i32 - borrow;
                if d < 0 {
                    d += 256;
                    borrow = 1;
                } else {
                    borrow = 0;
                }
                his[k] = d as u8;
            }
        }
        let mut wrongpub = pubk;
        wrongpub[0] ^= 0x01;
        let vt: [(&str, &[u8; 32], &[u8; 32], &[u8; 32], &[u8]); 8] = [
            ("valid", &h, &r, &s, &pubk),
            ("tamper", &hbad, &r, &s, &pubk),
            ("r0", &h, &zero, &s, &pubk),
            ("s0", &h, &r, &zero, &pubk),
            ("rN", &h, &CV_N, &s, &pubk),
            ("sN", &h, &r, &CV_N, &pubk),
            ("highS", &h, &r, &his, &pubk),
            ("wrongpk", &h, &r, &s, &wrongpub),
        ];
        line.clear();
        line.push_str(&format!("ver{}", i));
        for (nm, hh, rr, ss, pk) in vt.iter() {
            let v = secp_ecdsa_verify(hh, rr, ss, pk) as i32;
            line.push_str(&format!(" {}={}", nm, v));
            comb.update(&[v as u8]);
        }
        println!("{}", line);
    }

    // ── 5. tiny-key KAT: priv=1 ⇒ G ; priv=2 ⇒ 2G ──────────────────────────────
    {
        let mut p1 = [0u8; 32];
        p1[31] = 1;
        let pk1 = secp_pubkey(&p1).unwrap();
        line.clear();
        line.push_str("priv1_pub=");
        puthex(&mut line, &pk1);
        println!("{}", line);
        comb.update(&pk1);
        let mut p2 = [0u8; 32];
        p2[31] = 2;
        let pk2 = secp_pubkey(&p2).unwrap();
        line.clear();
        line.push_str("priv2_pub=");
        puthex(&mut line, &pk2);
        println!("{}", line);
        comb.update(&pk2);
    }

    // PRIMARY cross-language digest (sections 1–5).
    let cd = comb.finalize();
    line.clear();
    line.push_str("combined ");
    puthex(&mut line, &cd);
    println!("{}", line);

    // ── 6. end-to-end with the real curve ──────────────────────────────────────
    let mut e2e = Sha256::new();
    real_endtoend(&mut e2e);
    let ed = e2e.finalize();
    line.clear();
    line.push_str("combined_e2e ");
    puthex(&mut line, &ed);
    println!("{}", line);

    0
}

// ── raw-tx byte builders (mirror attrib.c e2e_rawtx / emit_push) ──────────────
fn put_varint(out: &mut Vec<u8>, n: u64) {
    if n < 0xfd {
        out.push(n as u8);
    } else if n <= 0xffff {
        out.push(0xfd);
        out.extend_from_slice(&(n as u16).to_le_bytes());
    } else if n <= 0xffff_ffff {
        out.push(0xfe);
        out.extend_from_slice(&(n as u32).to_le_bytes());
    } else {
        out.push(0xff);
        out.extend_from_slice(&n.to_le_bytes());
    }
}
fn emit_push(out: &mut Vec<u8>, data: &[u8]) {
    let l = data.len();
    if l < 76 {
        out.push(l as u8);
    } else if l < 256 {
        out.push(0x4c);
        out.push(l as u8);
    } else {
        out.push(0x4d);
        out.extend_from_slice(&(l as u16).to_le_bytes());
    }
    out.extend_from_slice(data);
}
// raw tx: version 1, 1 input (outpoint 0x11.., seq FFFFFFFF), 1 output (value 100000,
// scriptPubKey = OP_RETURN), locktime 0.
fn e2e_rawtx(ss: &[u8]) -> Vec<u8> {
    let mut raw = Vec::new();
    raw.extend_from_slice(&1u32.to_le_bytes());
    put_varint(&mut raw, 1);
    raw.extend_from_slice(&[0x11u8; 36]);
    put_varint(&mut raw, ss.len() as u64);
    raw.extend_from_slice(ss);
    raw.extend_from_slice(&0xFFFF_FFFFu32.to_le_bytes());
    put_varint(&mut raw, 1);
    raw.extend_from_slice(&100000u64.to_le_bytes());
    put_varint(&mut raw, 1);
    raw.push(0x6a);
    raw.extend_from_slice(&0u32.to_le_bytes());
    raw
}
// skeleton ParsedTx matching e2e_rawtx structure (empty scriptsig); used for sighash.
fn e2e_skeleton() -> ParsedTx {
    ParsedTx {
        version: 1,
        vin: vec![TxIn { prevout: [0x11u8; 36], scriptsig: Vec::new(), sequence: 0xFFFF_FFFF }],
        vout: vec![TxOut { value: 100000, spk: vec![0x6a] }],
        locktime: 0,
    }
}

fn emit_e2e(comb: &mut Sha256, name: &str, tx: &ParsedTx, k: usize) {
    let res = attribute(tx, k);
    let mut idh = String::new();
    puthex(&mut idh, &res.identity);
    println!("{} {}:{}", name, res.status, idh);
    comb.update(&[res.status]);
    comb.update(&res.sighash);
    comb.update(&res.identity);
}

fn real_endtoend(comb: &mut Sha256) {
    set_real_curve(true);

    // ── A. P2PKH, correctly signed ⇒ FOUND ─────────────────────────────────────
    {
        let mut priv_ = [0u8; 32];
        priv_[31] = 0x2A;
        let pubk = secp_pubkey(&priv_).unwrap();
        let h160 = hash160(&pubk);
        let mut sc = Vec::new();
        sc.extend_from_slice(&[0x76, 0xa9, 0x14]);
        sc.extend_from_slice(&h160);
        sc.extend_from_slice(&[0x88, 0xac]);
        let sk = e2e_skeleton();
        let sh = legacy_sighash(&sk, 0, &sc);
        let (r, s) = secp_ecdsa_sign(&priv_, &sh).unwrap();
        let der = der_sig(&r, &s);
        let mut ss = Vec::new();
        emit_push(&mut ss, &der);
        emit_push(&mut ss, &pubk);
        let raw = e2e_rawtx(&ss);
        match parse_tx(&raw) {
            Some(t) => emit_e2e(comb, "e2e_p2pkh_valid", &t, 0),
            None => println!("e2e_p2pkh_valid PARSEFAIL"),
        }
    }
    // ── B. P2PKH, signed by the WRONG key ⇒ verify-drop (status 2) ─────────────
    {
        let mut priv_ = [0u8; 32];
        priv_[31] = 0x2A;
        let mut wrong = [0u8; 32];
        wrong[31] = 0x2B;
        let pubk = secp_pubkey(&priv_).unwrap();
        let h160 = hash160(&pubk);
        let mut sc = Vec::new();
        sc.extend_from_slice(&[0x76, 0xa9, 0x14]);
        sc.extend_from_slice(&h160);
        sc.extend_from_slice(&[0x88, 0xac]);
        let sk = e2e_skeleton();
        let sh = legacy_sighash(&sk, 0, &sc);
        let (r, s) = secp_ecdsa_sign(&wrong, &sh).unwrap();
        let der = der_sig(&r, &s);
        let mut ss = Vec::new();
        emit_push(&mut ss, &der);
        emit_push(&mut ss, &pubk);
        let raw = e2e_rawtx(&ss);
        match parse_tx(&raw) {
            Some(t) => emit_e2e(comb, "e2e_p2pkh_wrongkey", &t, 0),
            None => println!("e2e_p2pkh_wrongkey PARSEFAIL"),
        }
    }
    // ── C. 2-of-2 P2SH multisig, two correct in-order sigs ⇒ FOUND ─────────────
    {
        let mut privs = [[0u8; 32]; 2];
        let mut pubs = [[0u8; 33]; 2];
        for i in 0..2 {
            privs[i][31] = 0x50 + i as u8;
            pubs[i] = secp_pubkey(&privs[i]).unwrap();
        }
        let mut rs = Vec::new();
        rs.push(0x52); // OP_2
        for i in 0..2 {
            rs.push(0x21);
            rs.extend_from_slice(&pubs[i]);
        }
        rs.push(0x52); // OP_2
        rs.push(0xae); // OP_CHECKMULTISIG
        let sk = e2e_skeleton();
        let sh = legacy_sighash(&sk, 0, &rs);
        let mut ss = Vec::new();
        ss.push(0x00); // NULLDUMMY
        for i in 0..2 {
            let (r, s) = secp_ecdsa_sign(&privs[i], &sh).unwrap();
            let der = der_sig(&r, &s);
            emit_push(&mut ss, &der);
        }
        emit_push(&mut ss, &rs);
        let raw = e2e_rawtx(&ss);
        match parse_tx(&raw) {
            Some(t) => emit_e2e(comb, "e2e_multisig_valid", &t, 0),
            None => println!("e2e_multisig_valid PARSEFAIL"),
        }
    }

    set_real_curve(false);
}
